File: src/ar_gpio.rs
/**
 * Copyright (C) 2018, Shmish <shmish90@gmail.com>
 *
 * This code is licensed under the MIT licence
 * found in the LICENCE file in the root directory
 * of this source tree.
 */

#[macro_use]
extern crate nix;

#[repr(C)]
pub struct gpiochip_info {
  pub name: [u8; 32],
  pub label: [u8; 32],
  pub lines: u32
}

#[repr(C)]
pub struct gpioline_info {
  pub line_offset: u32,
  pub flags: u32,
  pub name: [u8; 32],
  pub consumer: [u8; 32]
}

#[repr(C)]
pub struct gpiohandle_request {
  pub lineoffsets: [u32; 64],
  pub flags: u32,
  pub default_values: [u8; 64],
  pub consumer_label: [u8; 32],
  pub lines: u32,
  pub fd: i32
}

#[repr(C)]
pub struct gpiohandle_data {
  pub values: [u8; 64]
}

ioctl_read!(get_chip_info, 0xB4, 0x01, gpiochip_info);
ioctl_readwrite!(get_line_info, 0xB4, 0x02, gpioline_info);
ioctl_readwrite!(get_line_handle, 0xB4, 0x02, gpiohandle_request);
ioctl_readwrite!(get_line_values, 0xB4, 0x08, gpiohandle_data);
ioctl_readwrite!(set_line_values, 0xB4, 0x09, gpiohandle_data);


#[no_mangle]
pub extern fn gpiochip_create (fd: i32) -> *mut gpiochip_info {
  let mut info = gpiochip_info {
    name: [0; 32],
    label: [0; 32],
    lines: 0
  };

  unsafe {
    get_chip_info(fd, &mut info);
    std::mem::transmute(Box::new(info))
  }
}

#[no_mangle]
pub extern fn gpiochip_name (ptr: *mut gpiochip_info) -> *mut [u8; 32] {
  unsafe {
    let _gpiochip = &mut *ptr;
    std::mem::transmute(Box::new(_gpiochip.name))
  }
}

#[no_mangle]
pub extern fn gpiochip_label (ptr: *mut gpiochip_info) -> *mut [u8; 32] {
  unsafe {
    let _gpiochip = &mut *ptr;
    std::mem::transmute(Box::new(_gpiochip.label))
  }
}

#[no_mangle]
pub extern fn gpiochip_lines (ptr: *mut gpiochip_info) -> u32 {
  let _gpiochip = unsafe { &mut *ptr };
  _gpiochip.lines
}

#[no_mangle]
pub extern fn gpiochip_destroy (ptr: *mut gpiochip_info) {
  let _gpiochip: Box<gpiochip_info> = unsafe { std::mem::transmute(ptr) };
}


#[no_mangle]
pub extern fn gpioline_create (fd: i32) -> *mut gpioline_info {
  let mut info = gpioline_info {
    line_offset: 0,
    flags: 0,
    name: [0; 32],
    consumer: [0; 32]
  };

  unsafe {
    get_line_info(fd, &mut info);
    std::mem::transmute(Box::new(info))
  }
}

#[no_mangle]
pub extern fn gpioline_offset (ptr: *mut gpioline_info) -> u32 {
  let _gpioline = unsafe { &mut *ptr };
  _gpioline.line_offset
}

#[no_mangle]
pub extern fn gpioline_flags (ptr: *mut gpioline_info) -> u32 {
  let _gpioline = unsafe { &mut *ptr };
  _gpioline.flags
}

#[no_mangle]
pub extern fn gpioline_name (ptr: *mut gpioline_info) -> *mut [u8; 32] {
  unsafe {
    let _gpioline = &mut *ptr;
    std::mem::transmute(Box::new(_gpioline.name))
  }
}

#[no_mangle]
pub extern fn gpioline_consumer (ptr: *mut gpioline_info) -> *mut [u8; 32] {
  unsafe {
    let _gpioline = &mut *ptr;
    std::mem::transmute(Box::new(_gpioline.consumer))
  }
}

#[no_mangle]
pub extern fn gpioline_destroy (ptr: *mut gpioline_info) {
  let _gpioline: Box<gpioline_info> = unsafe { std::mem::transmute(ptr) };
}


#[no_mangle]
pub extern fn gpiohandle_request (fd: i32, line: u32, flags: u32) -> i32 {
  let mut request = gpiohandle_request {
    lineoffsets: [0; 64],
    flags: 0,
    default_values: [0; 64],
    consumer_label: [0; 32],
    lines: 0,
    fd: 0
  };

  request.lineoffsets[0] = line;
  request.flags = flags;
  request.default_values[0] = 0;
  request.consumer_label[0] = 64;
  request.lines = 1;

  unsafe { get_line_handle(fd, &mut request); }
  request.fd
}

#[no_mangle]
pub extern fn gpioline_set (fd: i32, value: u8) {
  let mut data = gpiohandle_data { values: [0; 64] };
  data.values[0] = value;
  unsafe { set_line_values(fd, &mut data); }
}

#[no_mangle]
pub extern fn gpioline_get (fd: i32) -> u8 {
  let mut data = gpiohandle_data { values: [0; 64] };
  unsafe { get_line_values(fd, &mut data); }
  data.values[0]
}

File: src/ar_gpio.js
/**
 * Copyright (C) 2018, Shmish <shmish90@gmail.com>
 *
 * This code is licensed under the MIT licence
 * found in the LICENCE file in the root directory
 * of this source tree.
 */

const { Library, ref } = require('fastcall')
const { openSync } = require('fs')

const ffi = new Library(`${__dirname}/ar_gpio.so`)
  .function('pointer gpiochip_create (int32)')
  .function('uint8* gpiochip_name (pointer)')
  .function('uint8* gpiochip_label (pointer)')
  .function('uint32 gpiochip_lines (pointer)')
  .function('void gpiochip_destroy (pointer)')

  .function('pointer gpioline_create (int32, uint32)')
  .function('uint32 gpioline_offset (pointer)')
  .function('uint32 gpioline_flags (pointer)')
  .function('uint8* gpioline_name (pointer)')
  .function('uint8* gpioline_consumer (pointer)')
  .function('void gpioline_destroy (pointer)')

  .function('int32 gpiohandle_request (int32, uint32, uint32)')
  .function('void gpioline_set (int32, uint8)')
  .function('uint8 gpioline_get (int32)')


const HIGH = Symbol.for('@Shmish/GPIO_HIGH')
const LOW = Symbol.for('@Shmish/GPIO_LOW')


class GPIOChipInfo {
  constructor (file) {
    const $ptr = ffi.interface.gpiochip_create(file)

    this.name = ref.readCString(ffi.interface.gpiochip_name($ptr))
    this.label = ref.readCString(ffi.interface.gpiochip_label($ptr))
    this.lines = ffi.interface.gpiochip_lines($ptr)

    ffi.interface.gpiochip_destroy($ptr)
    Object.freeze(this)
  }
}


class GPIOLineInfo {
  constructor (chip, line) {
    const $ptr = ffi.interface.gpioline_create(chip, line)

    this.offset = ffi.interface.gpioline_offset($ptr)
    this.flags = ffi.interface.gpioline_flags($ptr)
    this.name = ref.readCString(ffi.interface.gpioline_name($ptr))
    this.consumer = ref.readCString(ffi.interface.gpioline_consumer($ptr))

    ffi.interface.gpioline_destroy($ptr)
    Object.freeze(this)
  }
}


class GPIOChip {
  constructor (int) {
    this.fd = openSync(`/dev/gpiochip${int}`, 0o666)
    this.info = new GPIOChipInfo(this.fd)
  }

  get name () { return this.info.name }
  get label () { return this.info.label }
  get lines () { return this.info.lines }

  line (line, flags) {
    return new GPIOLine(
      ffi.interface.gpiohandle_request(this.fd, line, flags),
      this.fd,
      line
    )
  }
}


class GPIOLine {
  constructor (line_fd, chip_fd, line) {
    this.fd = line_fd
    this.info = new GPIOLineInfo(chip_fd, line)
  }

  get offset () { return this.info.offset }
  get flags () { return this.info.flags }
  get name () { return this.info.name }
  get consumer () { return this.info.consumer }

  ['get'] () {
    const state = ffi.interface.gpioline_get(this.fd)
    if (state === 1) return HIGH
    if (state === 0) return LOW
  }

  ['set'] (state) {
    if (state === HIGH || state === 1) ffi.interface.gpioline_set(this.fd, 1)
    if (state === LOW || state === 0) ffi.interface.gpioline_set(this.fd, 0)
  }
}


module.exports.GPIOChipInfo = GPIOChipInfo
module.exports.GPIOLineInfo = GPIOLineInfo
module.exports.GPIOChip = GPIOChip
module.exports.HIGH = HIGH
module.exports.LOW = LOW

module.exports.INPUT =       (1 << 0)
module.exports.OUTPUT =      (1 << 1)
module.exports.ACTIVE_LOW =  (1 << 2)
module.exports.OPEN_DRAIN =  (1 << 3)
module.exports.OPEN_SOURCE = (1 << 4)
